File: c/day02.c
#include <stdio.h>
#include <string.h>

#define SIZE 1024

/* available opcodes */
enum opcodes {
	OP_ADD = 1,
	OP_MUL = 2,
	OP_FIN = 99
};

int run_program(int *a)
{
	int i;
	const int oplen = 4;

	for(i=0;; i+=oplen) {
		switch(a[i]) {
			case OP_ADD:
				a[a[i+3]] = a[a[i+1]] + a[a[i+2]];
				break;
			case OP_MUL:
				a[a[i+3]] = a[a[i+1]] * a[a[i+2]];
				break;
			case OP_FIN:
				return 0;
			default:
				return -1;
		}
	}
}

int part1(int *a)
{
	a[1] = 12;
	a[2] = 2;

	run_program(a);

	return a[0];
}

int part2(int *a, int *b)
{
	int noun, verb;
	const int magic = 19690720;

	for(noun=0; noun<100; noun++) {
		for(verb=0; verb<100; verb++) {
			a[1] = noun;
			a[2] = verb;
			run_program(a);
			if(a[0] == magic) {
				return 100 * noun + verb;
			}
			/* restore initial state from backup */
			memcpy(a, b, SIZE * sizeof(int));
		}
	}
	return -1;
}


int main(int argc, char **argv)
{
	int i, m, a[SIZE], b[SIZE];

	i = 0;
	for(i=0; scanf("%d,", &m) != EOF; i++) {
		a[i] = m;
	}

	/* make backup of a in b */
	memcpy(b, a, SIZE * sizeof(int));

	printf("part1: %d\n", part1(a));
	
	/* restore backup for part 2 */
	memcpy(a, b, SIZE * sizeof(int));

	printf("part2: %d\n", part2(a, b));

	return 0;
}

File: go/day02.go
package main

import "fmt"

const (
    OP_ADD = 1
    OP_MUL = 2
    OP_FIN = 99
)

func run_program(a []int) int {
    const oplen = 4

    for i := 0; ; i += oplen {
        switch a[i] {
        case OP_ADD:
            a[a[i+3]] = a[a[i+1]] + a[a[i+2]]
        case OP_MUL:
            a[a[i+3]] = a[a[i+1]] * a[a[i+2]]
        case OP_FIN:
            return 0
        default:
            return -1
        }
    }
    return -1
}

func part1(a []int) int {
    a[1] = 12
    a[2] = 2

    run_program(a)
    return a[0]
}

func part2(a []int) int {
    const magic = 19690720

    b := make([]int, len(a))
    copy(b, a)
    for noun := 0; noun < 100; noun++ {
        for verb := 0; verb < 100; verb++ {
            a[1], a[2] = noun, verb
            run_program(a)
            if a[0] == magic {
                return 100 * noun + verb
            }
            copy(a, b)
        }
    }
    return -1
}

func main() {
    var m int
    var a []int

    for {
        n, err := fmt.Scanf("%d,", &m)
        if n == 0 || err != nil {
            break
        }
        a = append(a, m)
    }
    // store backup of a for pt 2
    b := make([]int, len(a))
    copy(b, a)

    fmt.Printf("part1: %d\n", part1(a))
    fmt.Printf("part2: %d\n", part2(b))
}
